File: skills/codex-cli-changelog/scripts/fetch_latest_release.py
#!/usr/bin/env python3
"""Fetch the latest usable OpenAI Codex release notes with GitHub CLI.

The Codex CLI repository publishes GitHub releases rather than a single
CHANGELOG.md. This helper uses `gh api` as the only GitHub access path. If
GitHub CLI auth is missing or expired, it stops and asks the user to log in.
"""
from __future__ import annotations

import argparse
import datetime as dt
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any


OWNER = "openai"
REPO = "codex"
SKILL_DIR = Path(__file__).resolve().parent.parent
DEFAULT_OUT_DIR = SKILL_DIR / "changelogs"

PRERELEASE_RE = re.compile(
    r"(?:^|[-_.])(alpha|beta|rc|pre|preview|nightly|snapshot|canary)(?:[-_.]|\d|$)",
    re.IGNORECASE,
)
VERSION_RE = re.compile(r"(?:codex[-_])?(?:rust[-_])?v?(\d+\.\d+\.\d+(?:\.\d+)?)", re.IGNORECASE)
GENERATED_SECTION_RE = re.compile(
    r"^\s*#{1,3}\s+(what'?s changed|new contributors|contributors|full changelog|changelog)\b",
    re.IGNORECASE,
)
FULL_CHANGELOG_RE = re.compile(r"^\s*(?:\*\*)?full changelog(?:\*\*)?\s*:", re.IGNORECASE)
PR_BULLET_RE = re.compile(r"^\s*[-*]\s+.*#\d{2,}.*@\S+", re.IGNORECASE)


class FetchError(RuntimeError):
    pass


def comparable_markdown(text: str) -> str:
    return "\n".join(
        line for line in text.replace("\r\n", "\n").replace("\r", "\n").splitlines()
        if not line.startswith("fetched:")
    ).strip()


def request_json(path: str, query: dict[str, str | int] | None = None) -> Any:
    endpoint = f"repos/{OWNER}/{REPO}{path}"
    if query:
        query_text = "&".join(f"{key}={value}" for key, value in query.items())
        endpoint = f"{endpoint}?{query_text}"
    args = ["gh", "api", endpoint]

    try:
        proc = subprocess.run(
            args,
            check=False,
            text=True,
            capture_output=True,
            timeout=60,
        )
    except FileNotFoundError as exc:
        raise FetchError("GitHub CLI `gh` is not installed or not on PATH.") from exc
    except subprocess.TimeoutExpired as exc:
        raise FetchError(f"`gh api {endpoint}` timed out.") from exc

    if proc.returncode != 0:
        stderr = proc.stderr.strip() or proc.stdout.strip()
        if "gh auth login" in stderr or "GH_TOKEN" in stderr or "Bad credentials" in stderr:
            raise FetchError(
                "GitHub CLI authentication is missing or expired. "
                "Run `gh auth login -h github.com` or set `GH_TOKEN`, then retry. "
                f"gh said: {stderr}"
            )
        raise FetchError(f"`{' '.join(args)}` failed: {stderr}")

    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError as exc:
        raise FetchError(f"`gh api {endpoint}` returned invalid JSON: {proc.stdout[:500]}") from exc


def release_label(release: dict[str, Any]) -> str:
    return " ".join(
        str(release.get(key) or "")
        for key in ("tag_name", "name", "html_url")
    ).strip()


def canonical_version(release: dict[str, Any]) -> str:
    label = release_label(release)
    match = VERSION_RE.search(label)
    if match:
        return match.group(1)
    return str(release.get("tag_name") or release.get("name") or release.get("id"))


def is_non_codex_release(release: dict[str, Any]) -> bool:
    label = release_label(release).lower()
    return "rusty-v8" in label or "librusty" in label


def is_prerelease_by_name(release: dict[str, Any]) -> bool:
    return bool(PRERELEASE_RE.search(release_label(release)))


def strip_generated_sections(body: str) -> tuple[str, list[str]]:
    lines = body.replace("\r\n", "\n").replace("\r", "\n").splitlines()
    kept: list[str] = []
    removed: list[str] = []
    skipping = False

    for line in lines:
        if GENERATED_SECTION_RE.match(line) or FULL_CHANGELOG_RE.match(line):
            removed.append(line.strip() or line)
            skipping = True
            continue
        if skipping:
            if re.match(r"^\s*#{1,3}\s+", line):
                # Resume only if this is not another generated section.
                if GENERATED_SECTION_RE.match(line) or FULL_CHANGELOG_RE.match(line):
                    removed.append(line.strip() or line)
                    continue
                skipping = False
                kept.append(line)
            else:
                removed.append(line.strip() or line)
            continue
        kept.append(line)

    cleaned = "\n".join(kept).strip()
    return cleaned, [item for item in removed if item]


def looks_like_generated_pr_notes(body: str) -> bool:
    lines = [line for line in body.splitlines() if line.strip()]
    if not lines:
        return True

    pr_bullets = [line for line in lines if PR_BULLET_RE.match(line)]
    generated_heading_count = sum(1 for line in lines if GENERATED_SECTION_RE.match(line))
    has_full_changelog = any(FULL_CHANGELOG_RE.match(line) for line in lines)

    if generated_heading_count or has_full_changelog:
        non_heading_lines = [
            line for line in lines
            if not GENERATED_SECTION_RE.match(line) and not FULL_CHANGELOG_RE.match(line)
        ]
        if non_heading_lines and len(pr_bullets) / len(non_heading_lines) >= 0.6:
            return True
    return len(pr_bullets) >= 8 and len(pr_bullets) / len(lines) >= 0.7


def usable_release(
    release: dict[str, Any],
    *,
    allow_generated: bool,
) -> tuple[bool, str, str, list[str]]:
    if release.get("draft"):
        return False, "", "draft release", []
    if release.get("prerelease") or is_prerelease_by_name(release):
        return False, "", "prerelease release", []
    if is_non_codex_release(release):
        return False, "", "non-Codex dependency release", []

    body = str(release.get("body") or "").strip()
    if not body:
        return False, "", "empty release body", []

    cleaned, removed = strip_generated_sections(body)
    if not allow_generated and looks_like_generated_pr_notes(body):
        if cleaned and not looks_like_generated_pr_notes(cleaned):
            return True, cleaned, "kept curated sections and removed generated PR notes", removed
        return False, "", "generated PR release notes only", removed

    if not cleaned:
        return False, "", "no content left after removing generated sections", removed
    return True, cleaned, "ok", removed


def find_latest_release(
    *,
    allow_generated: bool,
    max_pages: int,
) -> tuple[dict[str, Any], str, str, list[str]]:
    skipped: list[str] = []

    try:
        latest = request_json("/releases/latest")
        ok, body, reason, removed = usable_release(latest, allow_generated=allow_generated)
        if ok:
            return latest, body, reason, removed
        skipped.append(f"{latest.get('tag_name')}: {reason}")
    except FetchError as exc:
        raise FetchError(str(exc)) from exc

    for page in range(1, max_pages + 1):
        releases = request_json("/releases", {"per_page": 30, "page": page})
        if not releases:
            break
        for release in releases:
            ok, body, reason, removed = usable_release(release, allow_generated=allow_generated)
            if ok:
                if skipped:
                    release["_codex_changelog_skipped"] = skipped
                return release, body, reason, removed
            skipped.append(f"{release.get('tag_name')}: {reason}")

    raise FetchError("No usable latest Codex release found. Skipped: " + "; ".join(skipped[:20]))


def markdown_for_release(
    release: dict[str, Any],
    body: str,
    reason: str,
    removed: list[str],
) -> str:
    version = canonical_version(release)
    fetched = dt.date.today().isoformat()
    source = str(release.get("html_url") or f"https://github.com/{OWNER}/{REPO}/releases")
    tag = str(release.get("tag_name") or "")
    published = str(release.get("published_at") or "")

    frontmatter = [
        "---",
        f"version: {version}",
        f"tag: {json.dumps(tag)}",
        f"source: {source}",
        f"published: {json.dumps(published)}",
        f"fetched: {fetched}",
        "language: en",
        f"generated_notes_handling: {json.dumps(reason)}",
        "---",
        "",
        f"# Codex CLI {version}",
        "",
        body.strip(),
        "",
    ]
    return "\n".join(frontmatter)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--save", action="store_true", help="Save markdown to changelogs/<version>.md")
    ap.add_argument("--output-dir", default=str(DEFAULT_OUT_DIR), help="Directory for --save output")
    ap.add_argument("--json", action="store_true", help="Print selected release metadata as JSON")
    ap.add_argument(
        "--allow-generated",
        action="store_true",
        help="Allow auto-generated PR release notes if no curated content is present",
    )
    ap.add_argument("--max-pages", type=int, default=3, help="Release-list pages to scan after latest")
    args = ap.parse_args()
    out_dir = Path(args.output_dir).expanduser().resolve()

    try:
        release, body, reason, removed = find_latest_release(
            allow_generated=args.allow_generated,
            max_pages=args.max_pages,
        )
    except FetchError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 2

    version = canonical_version(release)
    markdown = markdown_for_release(release, body, reason, removed)

    if args.json:
        payload = {
            "version": version,
            "tag_name": release.get("tag_name"),
            "name": release.get("name"),
            "html_url": release.get("html_url"),
            "published_at": release.get("published_at"),
            "generated_notes_handling": reason,
            "removed_generated_lines": removed[:20],
            "skipped": release.get("_codex_changelog_skipped", []),
        }
        out_path = out_dir / f"{version}.md"
        if out_path.exists():
            existing = out_path.read_text(encoding="utf-8")
            payload["existing_local_path"] = str(out_path)
            payload["archive_status"] = (
                "unchanged"
                if comparable_markdown(existing) == comparable_markdown(markdown)
                else "different"
            )
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print(markdown)

    if args.save:
        out_dir.mkdir(parents=True, exist_ok=True)
        out_path = out_dir / f"{version}.md"
        if out_path.exists():
            existing = out_path.read_text(encoding="utf-8")
            if comparable_markdown(existing) == comparable_markdown(markdown):
                print(f"Skipped unchanged: {out_path}", file=sys.stderr)
                return 0
        out_path.write_text(markdown, encoding="utf-8")
        print(f"Saved: {out_path}", file=sys.stderr)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: skills/codex-cli-changelog/scripts/fetch_latest_release.ts
#!/usr/bin/env -S npx --yes tsx
/**
 * Fetch the latest usable OpenAI Codex release notes with GitHub CLI.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { execFileSync, spawnSync } from "node:child_process";

const OWNER = "openai";
const REPO = "codex";
const SKILL_NAME = "codex-cli-changelog";

const PRERELEASE_RE = /(?:^|[-_.])(alpha|beta|rc|pre|preview|nightly|snapshot|canary)(?:[-_.]|\d|$)/i;
const VERSION_RE = /(?:codex[-_])?(?:rust[-_])?v?(\d+\.\d+\.\d+(?:\.\d+)?)/i;
const GENERATED_SECTION_RE = /^\s*#{1,3}\s+(what'?s changed|new contributors|contributors|full changelog|changelog)\b/i;
const FULL_CHANGELOG_RE = /^\s*(?:\*\*)?full changelog(?:\*\*)?\s*:/i;
const PR_BULLET_RE = /^\s*[-*]\s+.*#\d{2,}.*@\S+/i;

interface Args {
  save: boolean;
  outputDir: string;
  json: boolean;
  allowGenerated: boolean;
  maxPages: number;
}

type Release = Record<string, unknown>;

class FetchError extends Error {}

function projectRoot(): string {
  try {
    const root = execFileSync("git", ["rev-parse", "--show-toplevel"], {
      cwd: process.cwd(),
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
      timeout: 5000,
    }).trim();
    return root ? path.resolve(root) : process.cwd();
  } catch {
    return process.cwd();
  }
}

function defaultRuntimeDir(): string {
  return path.join(projectRoot(), ".context", SKILL_NAME);
}

function defaultOutputDir(): string {
  return path.join(defaultRuntimeDir(), "changelogs");
}

function ensureRuntimeGitignore(): void {
  const runtimeDir = defaultRuntimeDir();
  fs.mkdirSync(runtimeDir, { recursive: true });
  const gitignore = path.join(runtimeDir, ".gitignore");
  if (fs.existsSync(gitignore)) return;
  fs.writeFileSync(
    gitignore,
    [
      "assets/rendered/",
      "assets/covers/",
      "*.png",
      "*.jpg",
      "*.jpeg",
      "*.webp",
      "*.gif",
      "*.mp4",
      "*.mov",
      ".DS_Store",
      "__pycache__/",
      "",
    ].join("\n"),
    "utf8",
  );
}

function isInside(child: string, parent: string): boolean {
  const relative = path.relative(parent, child);
  return relative === "" || (!!relative && !relative.startsWith("..") && !path.isAbsolute(relative));
}

function parseArgs(argv: string[]): Args {
  const args: Args = {
    save: false,
    outputDir: defaultOutputDir(),
    json: false,
    allowGenerated: false,
    maxPages: 3,
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--save") args.save = true;
    else if (arg === "--json") args.json = true;
    else if (arg === "--allow-generated") args.allowGenerated = true;
    else if (arg === "--max-pages") args.maxPages = Number.parseInt(argv[++i] ?? "3", 10);
    else if (arg === "--output-dir") args.outputDir = expandHome(argv[++i] ?? "");
    else if (arg === "--help" || arg === "-h") {
      printHelp();
      process.exit(0);
    } else {
      throw new FetchError(`Unknown argument: ${arg}`);
    }
  }
  return args;
}

function printHelp(): void {
  process.stdout.write(`Usage: fetch_latest_release.ts [--json] [--save] [--output-dir DIR] [--allow-generated] [--max-pages N]\n`);
}

function expandHome(value: string): string {
  if (value === "~") return process.env.HOME ?? value;
  if (value.startsWith("~/")) return path.join(process.env.HOME ?? "", value.slice(2));
  return value;
}

function today(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
}

function comparableMarkdown(text: string): string {
  return text
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .split("\n")
    .filter((line) => !line.startsWith("fetched:"))
    .join("\n")
    .trim();
}

function requestJson(apiPath: string, query?: Record<string, string | number>): unknown {
  let endpoint = `repos/${OWNER}/${REPO}${apiPath}`;
  if (query) endpoint += `?${new URLSearchParams(Object.entries(query).map(([k, v]) => [k, String(v)])).toString()}`;
  const proc = spawnSync("gh", ["api", endpoint], {
    encoding: "utf8",
    timeout: 60_000,
  });

  if (proc.error && (proc.error as NodeJS.ErrnoException).code === "ENOENT") {
    throw new FetchError("GitHub CLI `gh` is not installed or not on PATH.");
  }
  if (proc.error) throw new FetchError(`gh api ${endpoint} failed: ${proc.error.message}`);
  if (proc.status !== 0) {
    const stderr = (proc.stderr || proc.stdout || "").trim();
    if (stderr.includes("gh auth login") || stderr.includes("GH_TOKEN") || stderr.includes("Bad credentials")) {
      throw new FetchError(
        "GitHub CLI authentication is missing or expired. Run `gh auth login -h github.com` or set `GH_TOKEN`, then retry. " +
          `gh said: ${stderr}`,
      );
    }
    throw new FetchError(`gh api ${endpoint} failed: ${stderr}`);
  }

  try {
    return JSON.parse(proc.stdout);
  } catch {
    throw new FetchError(`gh api ${endpoint} returned invalid JSON: ${proc.stdout.slice(0, 500)}`);
  }
}

function str(value: unknown): string {
  return value == null ? "" : String(value);
}

function releaseLabel(release: Release): string {
  return [release.tag_name, release.name, release.html_url].map(str).join(" ").trim();
}

function canonicalVersion(release: Release): string {
  const match = releaseLabel(release).match(VERSION_RE);
  if (match) return match[1];
  return str(release.tag_name || release.name || release.id);
}

function isNonCodexRelease(release: Release): boolean {
  const label = releaseLabel(release).toLowerCase();
  return label.includes("rusty-v8") || label.includes("librusty");
}

function stripGeneratedSections(body: string): [string, string[]] {
  const kept: string[] = [];
  const removed: string[] = [];
  let skipping = false;

  for (const line of body.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n")) {
    if (GENERATED_SECTION_RE.test(line) || FULL_CHANGELOG_RE.test(line)) {
      removed.push(line.trim() || line);
      skipping = true;
      continue;
    }
    if (skipping) {
      if (/^\s*#{1,3}\s+/.test(line)) {
        if (GENERATED_SECTION_RE.test(line) || FULL_CHANGELOG_RE.test(line)) {
          removed.push(line.trim() || line);
          continue;
        }
        skipping = false;
        kept.push(line);
      } else {
        removed.push(line.trim() || line);
      }
      continue;
    }
    kept.push(line);
  }
  return [kept.join("\n").trim(), removed.filter(Boolean)];
}

function looksLikeGeneratedPrNotes(body: string): boolean {
  const lines = body.split("\n").filter((line) => line.trim());
  if (!lines.length) return true;
  const prBullets = lines.filter((line) => PR_BULLET_RE.test(line));
  const generatedHeadingCount = lines.filter((line) => GENERATED_SECTION_RE.test(line)).length;
  const hasFullChangelog = lines.some((line) => FULL_CHANGELOG_RE.test(line));
  if (generatedHeadingCount || hasFullChangelog) {
    const nonHeading = lines.filter((line) => !GENERATED_SECTION_RE.test(line) && !FULL_CHANGELOG_RE.test(line));
    if (nonHeading.length && prBullets.length / nonHeading.length >= 0.6) return true;
  }
  return prBullets.length >= 8 && prBullets.length / lines.length >= 0.7;
}

function usableRelease(release: Release, allowGenerated: boolean): [boolean, string, string, string[]] {
  if (release.draft) return [false, "", "draft release", []];
  if (release.prerelease || PRERELEASE_RE.test(releaseLabel(release))) return [false, "", "prerelease release", []];
  if (isNonCodexRelease(release)) return [false, "", "non-Codex dependency release", []];

  const body = str(release.body).trim();
  if (!body) return [false, "", "empty release body", []];

  const [cleaned, removed] = stripGeneratedSections(body);
  if (!allowGenerated && looksLikeGeneratedPrNotes(body)) {
    if (cleaned && !looksLikeGeneratedPrNotes(cleaned)) return [true, cleaned, "kept curated sections and removed generated PR notes", removed];
    return [false, "", "generated PR release notes only", removed];
  }
  if (!cleaned) return [false, "", "no content left after removing generated sections", removed];
  return [true, cleaned, "ok", removed];
}

function findLatestRelease(allowGenerated: boolean, maxPages: number): [Release, string, string, string[]] {
  const skipped: string[] = [];
  const latest = requestJson("/releases/latest") as Release;
  const latestUsable = usableRelease(latest, allowGenerated);
  if (latestUsable[0]) return [latest, latestUsable[1], latestUsable[2], latestUsable[3]];
  skipped.push(`${str(latest.tag_name)}: ${latestUsable[2]}`);

  for (let page = 1; page <= maxPages; page++) {
    const releases = requestJson("/releases", { per_page: 30, page }) as Release[];
    if (!Array.isArray(releases) || !releases.length) break;
    for (const release of releases) {
      const [ok, body, reason, removed] = usableRelease(release, allowGenerated);
      if (ok) {
        if (skipped.length) release._codex_changelog_skipped = skipped;
        return [release, body, reason, removed];
      }
      skipped.push(`${str(release.tag_name)}: ${reason}`);
    }
  }
  throw new FetchError(`No usable latest Codex release found. Skipped: ${skipped.slice(0, 20).join("; ")}`);
}

function markdownForRelease(release: Release, body: string, reason: string): string {
  const version = canonicalVersion(release);
  const source = str(release.html_url || `https://github.com/${OWNER}/${REPO}/releases`);
  const tag = str(release.tag_name);
  const published = str(release.published_at);
  return [
    "---",
    `version: ${version}`,
    `tag: ${JSON.stringify(tag)}`,
    `source: ${source}`,
    `published: ${JSON.stringify(published)}`,
    `fetched: ${today()}`,
    "language: en",
    `generated_notes_handling: ${JSON.stringify(reason)}`,
    "---",
    "",
    `# Codex CLI ${version}`,
    "",
    body.trim(),
    "",
  ].join("\n");
}

function main(): number {
  const args = parseArgs(process.argv.slice(2));
  const outDir = path.resolve(expandHome(args.outputDir));
  const [release, body, reason, removed] = findLatestRelease(args.allowGenerated, args.maxPages);
  const version = canonicalVersion(release);
  const markdown = markdownForRelease(release, body, reason);
  const outPath = path.join(outDir, `${version}.md`);

  if (args.json) {
    const payload: Record<string, unknown> = {
      version,
      tag_name: release.tag_name,
      name: release.name,
      html_url: release.html_url,
      published_at: release.published_at,
      generated_notes_handling: reason,
      removed_generated_lines: removed.slice(0, 20),
      skipped: release._codex_changelog_skipped ?? [],
    };
    if (fs.existsSync(outPath)) {
      const existing = fs.readFileSync(outPath, "utf8");
      payload.existing_local_path = outPath;
      payload.archive_status = comparableMarkdown(existing) === comparableMarkdown(markdown) ? "unchanged" : "different";
    }
    process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
  } else {
    process.stdout.write(markdown);
  }

  if (args.save) {
    if (isInside(outDir, defaultRuntimeDir())) ensureRuntimeGitignore();
    fs.mkdirSync(outDir, { recursive: true });
    if (fs.existsSync(outPath)) {
      const existing = fs.readFileSync(outPath, "utf8");
      if (comparableMarkdown(existing) === comparableMarkdown(markdown)) {
        process.stderr.write(`Skipped unchanged: ${outPath}\n`);
        return 0;
      }
    }
    fs.writeFileSync(outPath, markdown, "utf8");
    process.stderr.write(`Saved: ${outPath}\n`);
  }
  return 0;
}

try {
  process.exit(main());
} catch (error) {
  process.stderr.write(`ERROR: ${error instanceof Error ? error.message : String(error)}\n`);
  process.exit(2);
}
